### PR TITLE
[#232] Make face area magnitude projection-based

### LINKED ISSUE #232: Make face area magnitude robust for concave and star-shaped faces
Follow-up from #213.

`_ps_face_area_mag()` currently computes area by summing a triangle fan rooted at vertex 0 and taking magnitudes. That is only reliable for convex or fan-star-shaped faces. Concave or self-crossing/star faces can be mismeasured, which matters anywhere the helper is used for degeneracy decisions.

Suggested scope:

- Add focused tests for a concave face where the vertex-0 triangle fan overcounts or otherwise misjudges area.
- Replace `_ps_face_area_mag()` with a robust projected polygon-area calculation for planar simple faces, using the face frame normal/projection already available in core helpers.
- Keep self-crossing face policy explicit: either document that this helper reports signed-boundary absolute area, or leave self-crossing area handling to the segments/arrangement system.
- Verify existing cleanup/validation tests still pass.

### PR DESCRIPTION
**Pull request by susanw1**

### CODE REVIEW

**Comment by susanw1** [comment]
@codex review

File: src/polysymmetrica/core/funcs.scad [hunk 1]
```diff
@@ -1496,24 +1496,41 @@ function ps_face_frame_normal(verts, f, eps=1e-12) =
...
 //   verts = 3D vertex list
 //   f = face index loop
 function _ps_face_area_mag(verts, f) =
     (len(f) < 3) ? 0 :
-    ps_sum([
-        for (i = [1:1:len(f)-2])
-            norm(v_cross(verts[f[i]] - verts[f[0]], verts[f[i+1]] - verts[f[0]])) / 2
-    ]);
+    let(
+        n = ps_face_frame_normal(verts, f),
```
[the remark this answers is not in the document]
Comment (susanw1) [reply src/polysymmetrica/core/funcs.scad]: Fix is to stop needing a normalized face normal for area at all. A Newell/projected area vector gives the boundary area magnitude directly, with no fixed epsilon and no  first-three-vertices dependency. Changing _ps_face_area_mag() to use the area-vector magnitude. That keeps the concave fix while removing the scale-sensitive normal fallback.

[the remark this answers is not in the document]
Comment (susanw1) [reply src/polysymmetrica/core/funcs.scad]: @codex review - is that better?

File: src/polysymmetrica/core/funcs.scad [hunk 2]
```diff
@@ -1496,24 +1496,36 @@ function ps_face_frame_normal(verts, f, eps=1e-12) =
...
+    (axis == 0) ? (a[1] * b[2] - b[1] * a[2]) :
+    (axis == 1) ? (a[2] * b[0] - b[2] * a[0]) :
+                  (a[0] * b[1] - b[0] * a[1]);
+
+function _ps_face_area2_vector(verts, f) =
+    [
+        for (axis = [0:1:2])
+            ps_sum([
+                for (i = [0:1:len(f)-1])
+                    _ps_face_area_projected_term(verts[f[i]], verts[f[(i+1)%len(f)]], axis)
```
[the remark this answers is not in the document]
Comment (susanw1) [reply src/polysymmetrica/core/funcs.scad]: `_ps_face_area2_vector()` now subtracts `verts[f[0]]` before accumulating projected area terms, so the area calculation is translation-invariant again.

  Added regression tests for:

  - small collinear-first face area
  - unit square far from origin at [1e8, 1e8, 0]

[the remark this answers is not in the document]
Comment (susanw1) [reply src/polysymmetrica/core/funcs.scad]: @codex review again

File: src/polysymmetrica/core/funcs.scad [hunk 3]
```diff
@@ -1496,24 +1496,41 @@ function ps_face_frame_normal(verts, f, eps=1e-12) =
...
 // Arguments:
 //   verts = 3D vertex list
 //   f = face index loop
 function _ps_face_area_mag(verts, f) =
     (len(f) < 3) ? 0 :
-    ps_sum([
-        for (i = [1:1:len(f)-2])
-            norm(v_cross(verts[f[i]] - verts[f[0]], verts[f[i+1]] - verts[f[0]])) / 2
-    ]);
+    norm(_ps_face_area2_vector(verts, f)) / 2;
```
[the remark this answers is not in the document]
Comment (susanw1) [reply src/polysymmetrica/core/funcs.scad:1533]: What changed:

  - poly_cleanup(..., triangulate_nonplanar=true) now preserves non-planar unique-index loops through the first degenerate-face pass, so folded faces can reach triangulation.
  - The normal post-triangulation degenerate filter still runs, so bad triangles are still dropped.
  - Added a regression using the exact folded six-point loop from the review.

  Verified:

  - focused TestFuncs watcher run: pass
  - full src/tests/run_all.scad watcher run: pass
  - git diff --check: pass
  - openscad-docsgen core/models parse: pass

@codex review again?

[the remark this answers is not in the document]
Comment (susanw1) [reply src/polysymmetrica/core/funcs.scad:1533]: Yep, this is the same underlying distinction again: projected boundary area is the right planar-face measure, but it is not a safe degeneracy test for non-planar folded loops. I’ll extend the preservation rule to default cleanup too, while still letting the post-triangulation pass drop bad triangles when triangulation is requested.

[the remark this answers is not in the document]
Comment (susanw1) [reply src/polysymmetrica/core/funcs.scad:1533]: - Default poly_cleanup() now preserves unique-index non-planar folded loops instead of area-dropping them when the projected boundary vector cancels.
  - If triangulate_nonplanar=true, cleanup still applies the strict area filter after triangulation, so degenerate triangles are removed.
  - Added a regression for default cleanup preserving the folded six-point loop untriangulated.

  Verified:

  - focused TestFuncs watcher run: pass
  - full src/tests/run_all.scad watcher run: pass
  - git diff --check: pass
  - openscad-docsgen core/models parse: pass

@codex review

File: src/polysymmetrica/core/cleanup.scad [hunk 4]
```diff
@@ -22,15 +22,19 @@
...
+    preserve_nonplanar && !_ps_face_is_planar(verts, f, eps);
+
+function _ps_face_is_degenerate(verts, f, eps, preserve_nonplanar=false) =
     (len(f) < 3) ||
     (_ps_distinct_count(f) < 3) ||
     (_ps_distinct_count(f) != len(f)) ||
+    (!_ps_face_preserve_nonplanar_for_cleanup(verts, f, eps, preserve_nonplanar) &&
     // Area is quadratic in length, so compare against eps^2 (eps is linear tol).
-    (_ps_face_area_mag(verts, f) <= (eps * eps));
+    (_ps_face_area_mag(verts, f) <= (eps * eps)));
```
[the remark this answers is not in the document]
Comment (susanw1) [reply src/polysymmetrica/core/cleanup.scad]: Yes, “non-planar” alone is too broad because the fallback plane can misclassify fully collinear loops. I’ll refine the preservation gate so it only bypasses projected-area cleanup when the face has real fan triangle area.

[the remark this answers is not in the document]
Comment (susanw1) [reply src/polysymmetrica/core/cleanup.scad]: Change:

  - Non-planar preservation now also requires positive fan triangle area, so folded faces survive but fully collinear zero-area loops are still dropped.
  - Added a cube-plus-extra-collinear-face regression.

  Verified:

  - focused TestFuncs watcher run: pass
  - full src/tests/run_all.scad watcher run: pass
  - git diff --check: pass
  - openscad-docsgen core/models parse: pass

@codex review

File: src/polysymmetrica/core/cleanup.scad [hunk 5]
```diff
@@ -22,15 +22,28 @@
...
+    _ps_face_fan_area_mag(verts, f) > (eps * eps);
+
+function _ps_face_is_degenerate(verts, f, eps, preserve_nonplanar=false) =
     (len(f) < 3) ||
     (_ps_distinct_count(f) < 3) ||
     (_ps_distinct_count(f) != len(f)) ||
+    (!_ps_face_preserve_nonplanar_for_cleanup(verts, f, eps, preserve_nonplanar) &&
     // Area is quadratic in length, so compare against eps^2 (eps is linear tol).
-    (_ps_face_area_mag(verts, f) <= (eps * eps));
+    (_ps_face_area_mag(verts, f) <= (eps * eps)));
```
[the remark this answers is not in the document]
Comment (susanw1) [reply src/polysymmetrica/core/cleanup.scad]: This exposes the same pattern more cleanly: cleanup’s degeneracy test should not rely on signed projected boundary area alone. I’ll change the cleanup predicate to preserve any loop with either nonzero boundary area or nonzero fan triangle area, which keeps concave planar behavior while preserving folded and self-crossing loops and still dropping collinear loops.

[the remark this answers is not in the document]
Comment (susanw1) [reply src/polysymmetrica/core/cleanup.scad]: Change:

  - Cleanup degeneracy now treats a face as having area if either:
      - projected boundary area is positive, or
      - fan triangle area is positive.

  - This preserves planar self-crossing loops whose signed projected area cancels, while still dropping fully collinear zero-area loops.
  - Added a cube-backed bow-tie regression.

@codex review
